# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand

#' BagginsClass
#' 
#' BagginsClass：I am going on an adventure! this is an R6 object for rnaseq upstream workflow
#' @field root_dir character Path to the working directory for BAGGINS.
#' @field bamfiles_dir character Path to the directory containing aligned BAM files.
#' @field countfiles_dir character Path to the directory containing count files.
#' @field refer_genome character Path to the reference genome used for alignment.
#' @field gtf_file character Path to the annotation file in GTF format.
#' @field samtools_exe character Path to the SAMtools executable.
#' @field htseq_exe character Path to the HTSeq executable.
#' @field star_exe character Path to the STAR executable.
#' @field clean_str unnecessary str in the filenames
#' @return R6 object
#' 
#' @export
#' @examples
#' workflow <- BagginsClass$new(root_dir = ".")
BagginsClass <- R6::R6Class(
  "rnaseq_workflow",
  public = list(
    root_dir = NA,bamfiles_dir = NA,
    countfiles_dir= NA,refer_genome = NA,gtf_file = NA,
    samtools_exe = NA ,htseq_exe = NA ,star_exe = NA,
    clean_str = "Aligned.sortedByCoord.out.bam_counts.txt",
#' @param root_dir 根目录的路径
#' @param bamfiles_dir BAM文件目录的路径
#' @param countfiles_dir 计数文件目录的路径
#' @param refer_genome 参考基因组的路径
#' @param gtf_file GTF文件的路径
#' @param samtools_exe SAMtools可执行文件的路径
#' @param htseq_exe HTSeq可执行文件的路径
#' @param star_exe STAR可执行文件的路径
#' @param clean_str 清理字符串的值
    
    initialize = function(root_dir = NA,bamfiles_dir = NULL,
                          countfiles_dir = NULL,refer_genome = NULL,gtf_file = NULL,
                          samtools_exe = NULL ,htseq_exe = NULL ,star_exe = NULL,
                          clean_str = "Aligned.sortedByCoord.out.bam_counts.txt"){
      self$root_dir <- root_dir
      self$bamfiles_dir <- bamfiles_dir
      self$countfiles_dir<- countfiles_dir
      self$refer_genome <- refer_genome
      self$gtf_file <- gtf_file
      self$samtools_exe <- samtools_exe
      self$htseq_exe <- htseq_exe
      self$star_exe <- star_exe
      self$clean_str = clean_str
      self$info()
    },
#' @method info
#' 打印R6对象的信息
#' 
#' @return NULL
    info = function(){
      message("BagginsClass：I am going on an adventure! this is an R6 object for rnaseq upstream workflow ")
    },
#' @method initialize_dirs
#' 初始化R6对象的目录
#' 
#' @return NULL
    initialize_dirs = function(){
      rnaseq_placefile <- function(root_dir){
        files <- list.files(root_dir)
        file_ids <- rep(NA,length(files))
        for (i in 1:length(files)) {
          file_ids[i] <- stringr::str_split(files[i],"_")[[1]] %>% unlist()
        }
        
        for (file_id in file_ids){
          fs::dir_create(paste0(root_dir,"/",file_id))
          fs::dir_create(glue::glue("{root_dir}/bam_files"))
          fs::dir_create(glue::glue("{root_dir}/bam_files/{file_id}"))
        }
        df <- data.frame(id = file_ids,dir = files) %>% 
          dplyr::mutate(origin_dir = paste0(root_dir,"/",files),
                 dest_dir = paste0(root_dir,"/",id,"/",files),
          )
        for (i in 1:nrow(df)) {
          fs::file_move(df$origin_dir[i],df$dest_dir[i])
        }
      }
      print("info::preparing files and dirs for following use")
      fs::dir_create(self$bamfiles_dir)
      fs::dir_create(self$countfiles_dir)
      rnaseq_placefile(self$root_dir)
    },
#' @method fastq2matrix
#' 将fastq文件转换为矩阵
#' 
#' @return 转换后的矩阵
    fastq2matrix = function(){
      ## 将一组样本的R1和R2 RNA_seq一同对比，使用STAR软件
      rnaseq_star2annate <- function(star_exe,root_dir,refer_genome,bamfiles_dir){
        sample_list <- list.files(root_dir,"\\.fastq.gz",recursive = TRUE)
        for (sample in sample_list){
          input_dir = paste0(root_dir,sample)
          output_dir = paste0(bamfiles_dir,"/",sample)
          fastq_files = list.files(input_dir)
          fastq_r1 = paste0(input_dir,"/",fastq_files[1])
          fastq_r2 = paste0(input_dir,"/",fastq_files[2])
          system(paste(star_exe,
                       '--runThreadN', '8',
                       '--readFilesCommand', 'zcat',
                       '--quantMode', 'GeneCounts',
                       '--genomeDir', refer_genome,
                       '--readFilesIn', fastq_r1,fastq_r2,
                       '--twopassMode', 'Basic',
                       '--outSAMunmapped', 'None',
                       '--outSAMtype', 'BAM', 'SortedByCoordinate',
                       '--outFileNamePrefix', output_dir))
        }
      }
      rnaseq_bam4index <- function(samtool_exe,root_dir){
        bam_file_all <- list.files(root_dir,".bam",recursive = TRUE)
        for ( i in 1:length(bam_file_all)) {
          origin_path <- paste0(root_dir,'/',bam_file_all[i])
          print(paste("now we are working on",bam_file_all[i]))
          system(
            paste(samtool_exe,"index",origin_path))
        }
      }
      
      # 从rnaseq的bam文件进行未矫正的表达定量，使用HTSeq
      rnaseq_bam2counts <- function(htseq_exe,root_dir,gtf_dir,output_dir){
        # htseq-count -f bam -s no -t gene -i gene_id input.bam genes.gtf > counts.txt
        file_ids <- list.files(root_dir,".out.bam",recursive = TRUE)
        #file_lable <- rep(NA,length(file_ids))
        fs::dir_create(output_dir)
        for (i in 1:length(file_ids)){
          print(paste("now working on ",file_ids[i]))
          file_lable <- stringr::str_split(file_ids[i],"\\/")[[1]][2]
          #/home/rstudio/.local/bin/htseq-count -h
          system(
            paste(htseq_exe," -f bam -s no -t exon -i gene_id",
                  paste0("mds_expr/bam_files/",file_ids[i]),
                  gtf_dir,
                  ">",
                  paste0(output_dir,file_lable,"_counts.txt"))
          )
        }
      }
      rnaseq_count2matrix <- function(count_dir,clean_str = "Aligned.sortedByCoord.out.bam_counts.txt"){
        count_files <- list.files(count_dir,".txt")
        count_list <- list()
        for (i in 1:length(count_files)){
          count_lable <- stringr::str_remove(count_files[i],clean_str)
          count_table <- readr::read_tsv(paste0(count_dir,"/",count_files[i]),
                                         col_names = FALSE) %>% 
            dplyr::mutate(X2 = as.numeric(X2))
          colnames(count_table) <- c("GENEID",count_lable)
          count_list[[i]] <- count_table
          names(count_list)[i] <- count_lable
        }
        expr_df <- count_list[[1]] %>% 
          dplyr::left_join(count_list[[2]],by = "GENEID")
        for (i in 3:length(count_list)) {
          temp_df <- expr_df %>% 
            dplyr::left_join(count_list[[i]])
          expr_df <- temp_df
        }
        return(expr_df)
      }
      print("info::Step 1, refer the fastq files to genome")
      rnaseq_star2annate(self$star_exe,self$root_dir,self$refer_genome,self$bamfiles_dir)
      print("info::Step 2, index bam files and build count files in txt")
      rnaseq_bam4index(self$samtools_exe,self$bamfiles_dir)
      rnaseq_bam2counts(self$htseq_exe,self$bamfiles_dir,self$gtf_file,self$countfiles_dir)
      print("info::Step 3,build expression matrix ")
      df <- rnaseq_count2matrix(self$countfiles_dir,self$clean_str)
      return(df)
    }
    
  )
  
)
