# WARNING - Generated by {fusen} from dev/flat_first.Rmd: do not edit by hand

#' SamwiseClass
#' 
#' SamwiseClass:It's the job that's never started takes longest to finish.this is an R6 object for rrbs upstream workflow
#' @field root_dir character NULL 根目录的路径。
#' @field methyratio_dir character NULL 存储methyratio文件的目录的路径。
#' @field cpgmethy_dir character NULL 存储CpG甲基化文件的目录的路径。
#' @field cpgmethy_optimize character NULL 优化后的CpG甲基化文件的目录的路径。
#' @field promotermethy_optimize character NULL 优化后的启动子甲基化文件的目录的路径。
#' @field promotermethy_dir character NULL 存储启动子甲基化文件的目录的路径。
#' @field methyratio_py character NULL 用于计算methyratio的Python脚本的路径。
#' @field gene_fa character NULL 基因组序列的FASTA文件的路径。
#' @field bedtools_exe character NULL BEDTools可执行文件的路径。
#' @field cpg_bedfiles character NULL CpG岛的bed文件的路径。
#' @field promoter_bedfiles character NULL 启动子区域的bed文件的路径。
#' @field type character "promoter" 甲基化分析的类型（"cpg"或"promoter"）。
#' @field matrix_type character "Beta-value" 矩阵表示的类型（"Beta-value"或"M-value"）。
#' @field chrsex character "dame" 染色体性别（"dame"或"male"）。
#' 
#' @return R6 object
#' 
#' @export
#' @examples
#' object2 <- SamwiseClass$new(root_dir = "batch1117", type = "cpg")
SamwiseClass <- R6::R6Class(
  "rrbs_workflow", 
  public = list(
    root_dir = NULL, methyratio_dir = NULL, cpgmethy_dir = NULL,
    cpgmethy_optimize = NULL, promotermethy_optimize = NULL, promotermethy_dir = NULL, methyratio_py = NULL, gene_fa = NULL,
    bedtools_exe = NULL, cpg_bedfiles = NULL, promoter_bedfiles = NULL, type = "promoter", matrix_type = "Beta-value",
    chrsex = "dame", 
#' @method initialize
#' 初始化R6对象的构造函数。
#' @param root_dir character NULL 根目录的路径。
#' @param methyratio_dir character NULL 存储methyratio文件的目录的路径。
#' @param cpgmethy_dir character NULL 存储CpG甲基化文件的目录的路径。
#' @param cpgmethy_optimize character NULL 优化后的CpG甲基化文件的目录的路径。
#' @param promotermethy_optimize character NULL 优化后的启动子甲基化文件的目录的路径。
#' @param promotermethy_dir character NULL 存储启动子甲基化文件的目录的路径。
#' @param methyratio_py character NULL 用于计算methyratio的Python脚本的路径。
#' @param gene_fa character NULL 基因组序列的FASTA文件的路径。
#' @param bedtools_exe character NULL BEDTools可执行文件的路径。
#' @param cpg_bedfiles character NULL CpG岛的bed文件的路径。
#' @param promoter_bedfiles character NULL 启动子区域的bed文件的路径。
#' @param type character "promoter" 甲基化分析的类型（"cpg"或"promoter"）。
#' @param matrix_type character "Beta-value" 矩阵表示的类型（"Beta-value"或"M-value"）。
#' @param chrsex character "dame" 染色体性别（"dame"或"male"）。 
#' @return R6对象的实例。
    initialize = function(root_dir = NA, methyratio_dir = NA, 
                          cpgmethy_dir = NA, cpgmethy_optimize = NA,
                          promotermethy_optimize = NA, promotermethy_dir = NA, methyratio_py = "methratio.py",
                          gene_fa = "Genedata/hg19.fa", bedtools_exe = "bedtools", 
                          cpg_bedfiles ="Genedata/cpg_islands_hg19.bedgraph",
                          promoter_bedfiles = "Genedata/promoter_hg19_ensGene.bedgraph", type = "cpg", 
                          matrix_type = "Beta-value", chrsex = "dame") {
    self$root_dir <- root_dir
    self$methyratio_dir <- glue::glue("{root_dir}/bed_files")
    self$cpgmethy_dir <- glue::glue("{root_dir}/cpg_methy")
    self$cpgmethy_optimize <- glue::glue("{root_dir}/cpg_methy_optimize")
    self$promotermethy_optimize <- glue::glue("{root_dir}/promoter_methy_optimize")
    self$promotermethy_dir <- glue::glue("{root_dir}/promoter_methy")
    self$methyratio_py <- methyratio_py
    self$gene_fa <- gene_fa
    self$bedtools_exe <- bedtools_exe
    self$cpg_bedfiles <- cpg_bedfiles
    self$promoter_bedfiles <- promoter_bedfiles
    self$type <- type
    self$matrix_type <- matrix_type
    self$chrsex <- chrsex
    self$info()
  },
#' @method info
#' 打印R6对象的信息
#' 
#' @return str
  info = function() {
    message("SamwiseClass:It's the job that's never started takes longest to finish.this is an R6 object for rrbs upstream workflow")
  },
#' @method initialize_dirs
#' 初始化R6对象的目录
#' 
#' @return NULL

initialize_dirs = function() {
    print(paste("info:: Step 1,creat dirs for further use, as the following workflow after BSMSP-based procedure"))
    fs::dir_create(self$methyratio_dir)
    fs::dir_create(self$cpgmethy_dir)
    fs::dir_create(self$cpgmethy_optimize)
  }, 
#' @method bam2methyratio
#' 从BAM文件中计算methyratio值
#'
#' @return methyratio
bam2methyratio = function() {
    print(paste("info:: Step 2, Get methyratio values from bam files and transfer to bedgraph files"))
    bamfiles <- list.files(self$root_dir, pattern = "\\.bam$", recursive = FALSE)
    fs::dir_create(self$methyratio_dir)
    fs::dir_create(paste0(self$root_dir, "/temp"))
    fs::dir_create(paste0(self$root_dir, "/processed"))
    for (i in 1:length(bamfiles)) {
      print(paste("info::", paste0(i, "/", length(bamfiles)), ",start processing methyratio values", bamfiles[i]))
      output_dir_txt <- paste0(self$root_dir, "/temp/", "methyratio_", stringr::str_replace(bamfiles[i], ".bam",
        ".txt"))
      output_dir_bed <- paste0(self$methyratio_dir, "/", "methyratio_", stringr::str_replace(bamfiles[i], ".bam",
        ".bedgraph"))
      # python3 /home/rstudio/methy2expr/dev/BSMAPz-master/methratio.py -o test.txt
      #-d /home/rstudio/methy2expr/dev/hg19/hg19.fa 23T035358.bam
      system(paste("python2", self$methyratio_py, "-I -o", output_dir_txt, "-d", self$gene_fa, paste0(self$root_dir,
        "/", bamfiles[i])))
      print(paste("info::", paste0(i, "/", length(bamfiles)), ",start packing bedgraph files", bamfiles[i]))
      # 禁用科学计数法
      options(scipen = 9999)
      temp_df <- data.table::fread(output_dir_txt, header = TRUE, stringsAsFactors = FALSE) %>%
        dplyr::select(chr, pos, ratio) %>%
        dplyr::mutate(start = as.character(as.numeric(pos) - 1)) %>%
        dplyr::rename(end = pos) %>%
        dplyr::select(chr, start, end, ratio)
      data.table::fwrite(temp_df, file = output_dir_bed, sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
      print(paste("info::", paste0(i, "/", length(bamfiles)), ",start clean files in temp dirs with txt and bam"))
      fs::file_delete(output_dir_txt)
      fs::file_delete(paste0(self$root_dir, "/", bamfiles[i]))
      fs::file_move(paste0(self$root_dir, "/", stringr::str_replace(bamfiles[i], ".bam", ".tmpSrt.bam")), paste0(self$root_dir,
        "/processed/", stringr::str_replace(bamfiles[i], ".bam", ".tmpSrt.bam")))
      fs::file_move(paste0(self$root_dir, "/", stringr::str_replace(bamfiles[i], ".bam", ".tmpSrt.bam.bai")), paste0(self$root_dir,
        "/processed/", stringr::str_replace(bamfiles[i], ".bam", ".tmpSrt.bam.bai")))
      gc()
    }
  }, 
#' @method map2cpg4optimize
#' 将bedgraph文件映射到CpG岛或启动子区域并优化它们。
#'
#' @return 优化后的bedgraph文件。
  map2cpg4optimize = function() {
    print(paste("info:: Step 3, Get CpGs or promoter regions and optimize the bedgraph files"))
    rrbs_map2cpgislands <- function(bedtools_exe, root_dir, type = c("cpg", "promoter"), cpg_bedfiles, output) {
      bedfiles <- list.files(root_dir, ".bedgraph", recursive = TRUE)
      fs::dir_create(output)
      for (i in 1:length(bedfiles)) {
        print(paste(i, "/", length(bedfiles), ",we are now working on ", bedfiles[i]))
        # bedtools intersect -a methylation.bedgraph -b annotation.gtf -wa -wb > annotated_methylation.bed
        system(paste(bedtools_exe, "intersect -a", paste0(root_dir, "/", bedfiles[i]), "-b", cpg_bedfiles, "-wa -wb >",
          paste0(output, "/", stringr::str_replace(bedfiles[i], ".bedgraph", paste0("_", type, ".bedgraph")))))
      }
      gc()
    }
    # 优化跟cpg岛匹配后的bedgraph文件的序列和甲基化信号值
    rrbs_optimize4cpgislands <- function(root_dir, output) {
      bedfiles <- list.files(root_dir, ".bedgraph")
      fs::dir_create(output)
      for (i in 1:length(bedfiles)) {
        bedfile <- data.table::fread(paste0(root_dir, "/", bedfiles[i]))
        print(paste(paste0(i, "/", length(bedfiles)), ", now we are working on", bedfiles[i]))
        colnames(bedfile) <- c("chr_num", "start_chrn", "end_chrn", "methy_value", "chr", "start", "end", "CpGs")
        cpg_seq <- bedfile %>%
          dplyr::select(chr, start, end, CpGs) %>%
          dplyr::mutate(chr_cpgs = paste0(chr, "_", CpGs)) %>%
          dplyr::distinct(chr_cpgs, .keep_all = TRUE) %>%
          dplyr::select(start, end, chr_cpgs)
        bedfile2 <- bedfile %>%
          dplyr::group_by(CpGs, chr) %>%
          dplyr::summarise(methy_value = mean(methy_value)) %>%
          dplyr::mutate(chr_cpgs = paste0(chr, "_", CpGs)) %>%
          dplyr::left_join(cpg_seq, by = "chr_cpgs") %>%
          dplyr::select(chr, start, end, methy_value, CpGs) %>%
          dplyr::arrange(chr)
        data.table::fwrite(bedfile2, paste0(output, "/", bedfiles[i]), quote = FALSE)
      }
    }
    rrbs_optimize4promoter <- function(root_dir, output) {
      print(paste("info:: be careful, we will use at least 20GB RAM."))
      library(stringr)
      bedfiles <- list.files(root_dir, ".bedgraph")
      fs::dir_create(output)
      for (i in 1:length(bedfiles)) {
        bedfile <- data.table::fread(paste0(root_dir, "/", bedfiles[i]))
        print(paste(paste0(i, "/", length(bedfiles)), ", now we are working on", bedfiles[i]))
        colnames(bedfile) <- c("chr_num", "start_chrn", "end_chrn", "methy_value", "chr", "start", "end", "promoter",
          "unknow", "strand")
        bedfile <- bedfile %>%
          dplyr::mutate(promoter = str_split(promoter, "_")) %>%
          dplyr::mutate(promoter = sapply(promoter, `[`, 1))
        cpg_seq <- bedfile %>%
          dplyr::select(chr, start, end, promoter, strand) %>%
          dplyr::mutate(chr_cpgs = paste0(chr, "_", promoter, strand)) %>%
          dplyr::distinct(chr_cpgs, .keep_all = TRUE) %>%
          dplyr::select(chr, start, end, chr_cpgs)
        bedfile2 <- bedfile %>%
          dplyr::mutate(chr_cpgs = paste0(chr, "_", promoter, strand)) %>%
          dplyr::group_by(chr_cpgs) %>%
          dplyr::summarise(methy_value = mean(methy_value)) %>%
          dplyr::left_join(cpg_seq, by = "chr_cpgs") %>%
          dplyr::select(chr, start, end, methy_value, chr_cpgs) %>%
          dplyr::rename(promoter = chr_cpgs) %>%
          dplyr::arrange(chr)
        data.table::fwrite(bedfile2, paste0(output, "/", bedfiles[i]), quote = FALSE)
        gc()
      }
    }
    message(glue::glue("now we are on {self$type} mod."))
    if (self$type == "cpg") {
      rrbs_map2cpgislands(self$bedtools_exe, self$methyratio_dir, self$type, self$cpg_bedfiles, self$cpgmethy_dir)
      rrbs_optimize4cpgislands(self$cpgmethy_dir, self$cpgmethy_optimize)
      fs::dir_delete(self$cpgmethy_dir)
    } else {
      rrbs_map2cpgislands(self$bedtools_exe, self$methyratio_dir, self$type, self$promoter_bedfiles, self$promotermethy_dir)
      rrbs_optimize4promoter(self$promotermethy_dir, self$promotermethy_optimize)
      fs::dir_delete(self$promotermethy_dir)
    }
  }, 
#' @method pack2matrix
#' 将bedgraph文件打包为矩阵格式。
#'
#' @return 矩阵格式的数据框。
  pack2matrix = function() {

    rrbs_cpgBED2matrix <- function(root_dir, matrix_type = c("Beta-value", "M-value"), chrsex = "dame") {
      bedfiles <- list.files(root_dir, "\\.bedgraph", recursive = FALSE)
      union_cpgs <- list()
      for (i in 1:length(bedfiles)) {
        if (chrsex == "dame") {
          df_temp <- data.table::fread(paste0(root_dir, "/", bedfiles[i])) %>%
          dplyr::filter(chr != "chrX") %>%
          dplyr::filter(chr != "chrY") %>%
          dplyr::mutate(idx = paste0(chr, "~", start, "~", end, "~", CpGs))
          union_cpgs[[i]] <- df_temp$idx
        } else {
          df_temp <- data.table::fread(paste0(root_dir, "/", bedfiles[i])) %>%
          dplyr::mutate(idx = paste0(chr, "~", start, "~", end, "~", CpGs))
          union_cpgs[[i]] <- df_temp$idx
        }
      }
      union_cpg <- Reduce(intersect, union_cpgs)
      # 读取第一个bedgraph文件
      df_table <- data.table::fread(paste0(root_dir, "/", bedfiles[1])) %>%
        dplyr::filter(chr != "chrX") %>%
        dplyr::filter(chr != "chrY") %>%
        dplyr::mutate(idx = paste0(chr, "~", start, "~", end, "~", CpGs)) %>%
        dplyr::filter(idx %in% union_cpg) %>%
        dplyr::select(-CpGs)
      sample_id <- stringr::str_split(bedfiles[1], "_")[[1]][2]
      colnames(df_table) <- c("chr", "start", "end", sample_id, "idx")
      # 遍历剩余的bedgraph文件，并逐个进行合并
      for (i in 2:length(bedfiles)) {
        print(paste("working on item", i, "/", length(bedfiles)))
        temp_df <- data.table::fread(paste0(root_dir, "/", bedfiles[i])) %>%
          dplyr::filter(chr != "chrX") %>%
          dplyr::filter(chr != "chrY") %>%
          dplyr::mutate(idx = paste0(chr, "~", start, "~", end, "~", CpGs)) %>%
          dplyr::filter(idx %in% union_cpg) %>%
          dplyr::select(-all_of(c("chr", "start", "end", "CpGs")))
        sample_id <- stringr::str_split(bedfiles[i], "_")[[1]][2]
        colnames(temp_df) <- c(sample_id, "idx")
        df_table <- dplyr::left_join(df_table, temp_df, by = "idx")
      }

      df_table_idx <- df_table %>%
        dplyr::select(chr, start, end, idx)
      # 将df_table_idx写入文件
      data.table::fwrite(df_table_idx, "df_cpg_idx.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
      # 重新整理df_table，移动idx列至最前面
      df_table <- df_table %>%
        dplyr::select(-all_of(c("chr", "start", "end"))) %>%
        dplyr::relocate(idx)
      # 根据matrix_type选择处理方式
      if (matrix_type == "M-value") {
        # 计算M-value
        matrixdf <- as.matrix(df_table[, -1])
        result <- log2(matrixdf/(1 - matrixdf))
        df <- as.data.frame(result) %>%
          dplyr::mutate(idx = df_table$idx) %>%
          dplyr::relocate(idx)
        colnames(df) <- colnames(df_table)
      } else {
        df <- df_table
      }
      # 将df保存为RDS文件
      saveRDS(df, "methylationSet_cpgs.RDS")
      return(df)
    }
    rrbs_promoterBED2matrix <- function(root_dir, matrix_type = c("Beta-value", "M-value"), chrsex = "dame") {
      bedfiles <- list.files(root_dir, "\\.bedgraph", recursive = FALSE)
      union_cpgs <- list()
      for (i in 1:length(bedfiles)) {
        if (chrsex == "dame") {
          df_temp <- data.table::fread(paste0(root_dir, "/", bedfiles[i])) %>%
          dplyr::filter(chr != "chrX") %>%
          dplyr::filter(chr != "chrY")
          dplyr::union_cpgs[[i]] <- df_temp$promoter
        } else {
          df_temp <- data.table::fread(paste0(root_dir, "/", bedfiles[i]))
          union_cpgs[[i]] <- df_temp$promoter
        }
      }
      union_cpg <- Reduce(intersect, union_cpgs)
      # 读取第一个bedgraph文件
      df_table <- data.table::fread(paste0(root_dir, "/", bedfiles[1])) %>%
        dplyr::filter(chr != "chrX") %>%
        dplyr::filter(chr != "chrY") %>%
        dplyr::filter(promoter %in% union_cpg) %>%
        dplyr::rename(dix = promoter)
      sample_id <- stringr::str_split(bedfiles[1], "_")[[1]][2]
      colnames(df_table) <- c("chr", "start", "end", sample_id, "idx")
      # 遍历剩余的bedgraph文件，并逐个进行合并
      for (i in 2:length(bedfiles)) {
        print(paste("working on item", i, "/", length(bedfiles)))
        temp_df <- data.table::fread(paste0(root_dir, "/", bedfiles[i])) %>%
          dplyr::filter(chr != "chrX") %>%
          dplyr::filter(chr != "chrY") %>%
          dplyr::filter(promoter %in% union_cpg) %>%
          dplyr::rename(dix = promoter) %>%
          dplyr::select(-all_of(c("chr", "start", "end")))
        sample_id <- stringr::str_split(bedfiles[i], "_")[[1]][2]
        colnames(temp_df) <- c(sample_id, "idx")
        df_table <- left_join(df_table, temp_df, by = "idx")
      }

      df_table_idx <- df_table %>%
        dplyr::select(chr, start, end, idx)
      # 将df_table_idx写入文件
      data.table::fwrite(df_table_idx, "df_promoter_idx.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)
      # 重新整理df_table，移动idx列至最前面
      df_table <- df_table %>%
        dplyr::select(-all_of(c("chr", "start", "end"))) %>%
        dplyr::relocate(idx)
      # 根据matrix_type选择处理方式
      if (matrix_type == "M-value") {
        # 计算M-value
        matrixdf <- as.matrix(df_table[, -1])
        result <- log2(matrixdf/(1 - matrixdf))
        df <- as.data.frame(result) %>%
          dplyr::mutate(idx = df_table$idx) %>%
          dplyr::relocate(idx)
        colnames(df) <- colnames(df_table)
      } else {
        df <- df_table
      }
      # 将df保存为RDS文件
      saveRDS(df, "methylationSet_promoter.RDS")
      return(df)
    }
    print(paste("info:: Step 4, Get methylation matrix among samples && calculate M-value in option"))
    if (self$type == "cpg") {
      df <- rrbs_cpgBED2matrix(self$cpgmethy_optimize, self$matrix_type, self$chrsex)
    } else {
      df <- rrbs_promoterBED2matrix(self$promotermethy_optimize, self$matrix_type, self$chrsex)
    }
    print(paste("info:: congratuations! You have successfully build a methylation matrix.You can find the methylation matrix in RDS format in ",
      getwd(), "/methylationSet.RDS"))
    return(df)
  }))
